python/mspasspy/db/database.py

```diff
@@ -308,53 +375,726 @@ def save_data(self, mspass_object, storage_mode='gridfs', dfile=None, dir=None,
+        not work.   Impossible conversions will not abort the function but 
+        will generate an error message printed to stdout.  The function 
+        continues on so if there are a large number of such errors the 
+        output could become voluminous.  ALWAYS run dbverify before trying 
+        this function (directly or indirectly through the command line 
+        tool dbclean).   
+        
+        :param db:  Database handle to be updated
+        :param collection:  MongoDB collection to be updated
+        :param query: optional query string passed to find database 
+        collection method.  Can be used to limit edits to documents 
+        matching the query.  Default is the entire collection.
+        :param verbose:  when true edit will produce one or more lines of 
+        printed output for each change it makes.  The default is false.
+        Needless verbose should be avoided unless you are certain the 
+        number of changes it will make are small.  
+        """
+        dbcol=self[collection]
+        schema=self.database_schema
+        col_schema=schema[collection]
+        counts=dict()
+        cursor=dbcol.find(query)
+        for doc in cursor:
+            n=0
+            id=doc.pop('_id')
+            if verbose:
+                print("////////Document id=",id,'/////////')
+            up_d=dict()
+            for k in doc:
+                val=doc[k]
+                if not col_schema.is_defined(k):
+                    if verbose:
+                        print('Warning:  in doc with id=',id,
+                            'found key=',k,' that is not defined in the schema')
+                        print('Value of key-value pair=',val)
+                        print('Cannot check type for an unknown attribute name')
+                    continue
+                if not isinstance(val,col_schema.type(k)):
+                    try:
+                        newval=col_schema.type(k)(val)
+                        up_d[k]=newval
+                        if verbose:
+                            print('Changed data for key=',k,' from ',val,' to ',newval)
+                        if k in counts:
+                            counts[k]+=1
+                        else:
+                            counts[k]=1
+                        n+=1
+                    except Exception as err:
+                        print("////////Document id=",id,'/////////')
+                        print('WARNING:  could not convert attribute with key=',
+                            k,' and value=',val,' to required type=',
+                            col_schema.type(k))
+                        print('This error was thrown and handled:  ')
+                        print(err)
+            
+            if n>0:
+                dbcol.update_one({'_id' : id},{'$set' : up_d})
+                
+        return counts
+
+    def check_links(self, normalize='site', wf="wf_TimeSeries", wfquery={}, verbose=False, error_limit=1000):
```

Comment:
Maybe these newly added method should be defined private at least for now? I can see the usage of them are very limited which makes it confusing if we open them to the users.